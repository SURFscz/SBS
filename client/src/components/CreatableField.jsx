import React from "react";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {Tooltip} from "@surfnet/sds";
import "./CreatableField.scss";
import {isEmpty, stopEvent} from "../utils/Utils";

export default function CreatableField({
                                           onChange,
                                           name,
                                           value,
                                           values,
                                           addValue,
                                           removeValue,
                                           toolTip = null,
                                           placeHolder = null,
                                           error = false,
                                           disabled = false
                                       }) {

    return (
        <div className={`creatable-field ${error ? "error" : ""}`}>
            <label htmlFor={name}>{name}
                {toolTip && <Tooltip tip={toolTip}/>}
            </label>
            <div className={`inner-creatable-field ${error ? "error" : ""}${disabled ? "disabled" : ""}`}>
                {values.map(val =>
                    <div key={val.name} className="creatable-tag">
                        <span>{val.name}</span>
                        {!disabled && <span onClick={removeValue(val)}>
                                            <FontAwesomeIcon icon="times"/>
                                        </span>}
                        {disabled && <span></span>}

                    </div>)}
                {!disabled && <textarea id="creatable-field" value={value} onChange={onChange} onBlur={addValue}
                                        onKeyDown={e => {
                                            if (e.key === "Enter" || e.key === " " || e.key === "Spacebar") {
                                                addValue(e);
                                                setTimeout(() => document.getElementById("creatable-field").focus(), 50);
                                                return stopEvent(e);
                                            } else if (e.key === "Backspace" && isEmpty(value) && values.length > 0) {
                                                const val = values[values.length - 1];
                                                removeValue(val)();
                                            } else if (e.key === "Tab") {
                                                addValue(e);
                                            }
                                        }}
                                        placeholder={placeHolder} cols={1}/>}
            </div>
        </div>
    );
}
