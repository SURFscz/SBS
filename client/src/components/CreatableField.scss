@import "../stylesheets/vars.scss";

div.creatable-field {
    display: flex;
    position: relative;
    flex-direction: column;

    label {
        display: inline-block;
        margin: 12px 0 6px 0;
        min-width: 210px;
        color: black;
        position: relative;
        font-weight: bold;
    }

    span.tool-tip-section {
        font-size: 14px;
        color: $lighter-grey;
        cursor: pointer;
        padding: 0 4px;

        &:hover {
            color: darken($light-blue, 20%);
        }
    }

    .inner-creatable-field {
        display: flex;
        flex-wrap: wrap;
        width: 100%;
        border: 1px solid $dark-grey;
        border-radius: $br;
        font-size: 16px;

        &:focus-within {
            border: 2px solid #015fcb;
        }

        &.error {
            background-color: $error-red-background;
            border: 1px solid $error-red;
        }

    }

    textarea {
        border: none;
        flex-grow: 2;
        display: inline-block;
        background-color: transparent;
        font-size: 16px;
        margin: 14px 5px 5px 5px;

        &:focus {
            border: none;
            outline: none;
        }
    }

    div.creatable-tag {
        border: 1px solid $lighter-grey;
        border-radius: $br;
        margin: 10px 5px;
        display: inline-block;
        position: relative;
        background-color: $lightest-grey;
        padding: 4px 26px 4px 10px;
        max-height: 32px;

        span {
            &:first-child {
                word-break: break-all;
            }

            &:last-child {
                padding: 8px 14px;
                cursor: pointer;
                font-size: 14px;
                color: $dark-grey;
                position: absolute;
                right: -8px;
                top: -2px;

                &.disabled {
                    cursor: not-allowed;
                    color: $green;
                }
            }
        }
    }


}