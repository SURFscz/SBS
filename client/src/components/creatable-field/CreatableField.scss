@import "../../stylesheets/vars";

div.creatable-field {
    display: flex;
    position: relative;
    flex-direction: column;

    label {
        display: inline-block;
        margin: 12px 0 6px 0;
        min-width: 210px;
        color: black;
        position: relative;
        font-weight: 600;
    }

    .inner-creatable-field {
        display: flex;
        flex-wrap: wrap;
        width: 100%;
        border: 1px solid $dark-grey;
        border-radius: $br;
        font-size: 16px;

        &:focus-within {
            border: 2px solid #015fcb;
        }

        &.error {
            background-color: $error-red-background;
            border: 1px solid $error-red;
        }

        &.disabled {
            cursor: not-allowed;
            background-color: $lightest-grey;
            min-height: 40px;
        }

    }

    textarea {
        border: none;
        flex-grow: 2;
        display: inline-block;
        background-color: transparent;
        font-size: 16px;
        padding: 12px 5px 0 6px;
        line-height: 16px;

        &:focus {
            border: none;
            outline: none;
        }
    }

    div.creatable-tag {
        border: 1px solid $lighter-grey;
        border-radius: $br;
        margin: 5px;
        display: inline-block;
        position: relative;
        background-color: $lightest-grey;
        padding: 4px 26px 4px 10px;
        max-height: 32px;

        span {
            &:first-child {
                word-break: break-all;
            }

            &:last-child {
                padding: 8px 14px;
                cursor: pointer;
                font-size: 14px;
                color: $dark-grey;
                position: absolute;
                right: -8px;
                top: -2px;

                &.disabled {
                    cursor: not-allowed;
                    color: $green;
                }
            }
        }
    }


}
