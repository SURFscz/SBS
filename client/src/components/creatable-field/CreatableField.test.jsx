import React from "react";
import { render } from "@testing-library/react";
import CreatableField from "./CreatableField";

jest.mock("../../locale/I18n", () => ({
  t: key => key
}));

const options = [
  { value: "one", label: "One" },
  { value: "two", label: "Two" }
];

describe("CreatableField", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(
      <CreatableField
        value={options[0]}
        onChange={jest.fn()}
        options={options}
        placeholder="Select or create"
        isMulti={false}
        isClearable={true}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
