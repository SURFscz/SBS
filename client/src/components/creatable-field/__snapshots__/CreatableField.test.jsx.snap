// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatableField matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="creatable-field "
  >
    <label />
    <div
      class="inner-creatable-field "
    >
      <textarea
        cols="1"
        id="creatable-field"
      >
        [object Object]
      </textarea>
    </div>
  </div>
</DocumentFragment>
`;
